04-onion-routing.md

```diff
@@ -292,7 +292,7 @@ The writer:
     - otherwise:
       - MUST set `total_msat` to the amount it wishes to pay.
     - MUST ensure that the total `amount_msat` of the HTLC set which arrives at the payee
-      is equal to `total_msat`.
+      is equal to or greater than `total_msat`.
```

Comment:
I don't think this change is necessary anymore when make the change to allow `htlc_amt > amt_to_fwd` in the later commit?

Reply:
I think we have to keep it, don't we? Because if we overpay one HTLC in the set, that means that the total HTLC set will overpay the `total_amount`?

Reply:
> I don't think so, because you only count amt_to_forward towards total_msat. The overpayment is already absorbed at the htlc level.

I see what you mean, that depends on what you use to compute your "HTLC set amount". Currently the spec does it using `amount_msat`, which is the value from `update_add_htlc`: in that case, we do need to explicitly allow over-paying.

What you suggest is that we should define the "HTLC set amount" to be the sum of the onion's `amt_to_fwd`, in which case we don't necessarily need to allow over-paying.

That would be a bit awkward for the scenario highlighted in #1031 though. Let's imagine that you retry a partial payment of `500 msat` and are forced to pay `510 msat` to satisfy `min_htlc` constraints in the route. With your proposal, the sender would be forced to put `500 msat` in the onion `amt_to_fwd`, which means giving the additional fee to intermediate nodes instead of the final node (and feels a bit weird, doesn't it?). Whereas if we allow over-paying the `total_amount_msat`, the sender can decide whether they want to give that extra fee to intermediate nodes or to the final node.

I understand your point of view, but my current feeling is to go for the extra flexibility of allowing the `total_amount_msat` over-payment. Let's see what others prefer, I can change that if that's the consensus.

Reply:
>Let's imagine that you retry a partial payment of 500 msat and are forced to pay 510 msat to satisfy min_htlc constraints in the route. With your proposal, the sender would be forced to put 500 msat in the onion amt_to_fwd, which means giving the additional fee to intermediate nodes instead of the final node.

I don't think you are giving it to the intermediate node straight away? The instruction is still to forward 510 sat to the final node. Only if the intermediate node would first try to pay lower amounts, they could steal some of those sats.

Also I think this only applies to the final shard. All the other shards would just put `510 msat` in the onion `amt_to_fwd`? It satisfies `min_htlc` and makes sure that a maximum of sats is delivered to the final destination. The overshoot can be compensated in the final shard to end up with the exact pre-determined `total_msat`.

What remains then is a sender being forced to overpay on the final shard because of a `min_htlc` constraint and the possibility of a routing node trying to steal the difference. It wouldn't increase the total cost of the payment though, because all prior shards were already 'optimal' in that they don't allow any fee stealing. The difference is just in where that final overpayment ends up IF the routing node decides to probe.

Even if we fix this for the final hop, it is still possible to shift the fee distribution to the intermediate hops.

If senders would always assume that overpayment of `total_msat` is acceptable, there wouldn't be a way anymore to reject such payments. You can fail the htlc set if it overpays, but the sender wouldn't know what's wrong. As mentioned before, overpayment isn't always desirable and can complicate accounting.

Reply:
I thought the probing attack was actually fixed by allowing nodes to forward less (if the amount was inflated because of `min_htlc`), just as is allowed currently for intermediate nodes.

Can you otherwise give a very concrete example, to make sure we're really talking about the same thing?

Reply:
How can you not allow nodes to do that? They can forward whatever they want.

Getting back to an example to clarify things: I think the following table lists the possibilities to deliver 5 sat to a node through a channel that has a min_htlc routing policy of 10 sat.

| amt_to_fwd A | htlc_amt A->B | amt_to_fwd B | current behavior routing node | current behavior final node | relaxed htlc_amt check final node |
| ------------ | ------------- | ------------ | ----------------------------- | --------------------------- | --------------------------------- |
| 10           | 9             | 10           | REJECT                        | REJECT                      | REJECT                            |
| 10           | 10            | 10           | ACCEPT                        | ACCEPT                      | ACCEPT                            |
| 10           | 11            | 10           | ACCEPT                        | REJECT                      | ACCEPT                            |
| 10           | 9             | 5            | ACCEPT                        | REJECT                      | ACCEPT                            |
| 10           | 10            | 5            | ACCEPT                        | REJECT                      | ACCEPT                            |
| 10           | 11            | 5            | ACCEPT                        | REJECT                      | ACCEPT                            |

To me it seems that the relaxed `htlc_amt` check alone fixes the probing issue because it makes the behavior of final nodes and routing nodes the same.

Reply:
> To me it seems that the relaxed htlc_amt check alone fixes the probing issue because it makes the behavior of final nodes and routing nodes the same.

I think that's what you're missing, this is incorrect and the three last rows in your table are prone to probing. The onion's `amt_to_fwd` _must be the same for the next-to-last node and the last node_ to prevent probing. The underlying reason is that _intermediate nodes will never let you underpay_ (because that wouldn't satisfy their fee requirement), while underpaying the recipient becomes possible if its onion's `amt_to_fwd` is smaller than the next-to-last node's. So intermediate nodes can always try to underpay, and if that's accepted, it means the next node is the recipient. Does that make sense? Or should I detail a complete example (which is a bit tedious, since we need to have two hops in the table to correctly explain this).

Reply:
Ah I get it now. This is all about setting `amt_to_fwd` (Matt mentioned this too in https://github.com/lightning/bolts/pull/1032#discussion_r992442477). Sender's must make sure that the whole path is 'tight' so that no underpayment can happen anywhere. Because of the relaxation of `amt_to_fwd`, the 1 msat overpayment will be possible for both routing and intermediate nodes which fixes the privacy issue.

The next problem arises when a sender needs to overpay because of `min_htlc` and overshoots `total_msat`. They could decide not to adhere to the final amt_to_forward setting requirement above, but would sacrifice privacy. Therefore the second change to allow overshooting `total_msat` instead.

One thing to note is that there may be an additional cost for senders to keeping the path tight. If there is an intermediate node with a high `min_htlc`, the sender may not want to carry over that high amount to all channels further downstream. For those channels a proportional fee may be charged, in which case it would be cheaper for the sender to reduce amt_to_fwd right after the `min_htlc` has been passed. If the sender decides to do so, it becomes possible again to identify a node as being non-final.

For inbound fees though, I am still not sure how to handle them with these changes. An htlc comes in, but I can only decide that not enough inbound fees were paid until the mpp set is complete. All htlcs need to be held until then, and fresh channel updates need to be returned for all of them. It also looks like a layering violation, as fees pertain to the htlc level only.

Reply:
Actually, for inbound fees there may not be a problem. In think the sender can safely set the final `amt_to_fwd` to a lower value than the second-last `amt_to_fwd` _if_ the difference is equal to the inbound fee that is required. If a routing node would try to underpay (and identify the final node), the htlc would still fail because the final node requires `htlc_amt = amt_to_fwd + inbound_fee`.

Reply:
One more remark here: if a node recently lowered its (regular outbound) fee and the sender isn't aware of that yet, a window opens up momentarily where probing could reveal some information. Maybe something to keep in mind for payments with high privacy requirements? I think that to fully seal a the route, that `required_incoming_htlc_amt` field is needed still.

Reply:
> Sender's must make sure that the whole path is 'tight' so that no underpayment can happen anywhere. 

Exactly. That is most likely the default behavior of senders, since they build the onion based on the latest channel updates they received and have no reason to overpay intermediate fees. That's why I believe the simple changes in this PR are a good enough mitigation of that probing vector.

> if a node recently lowered its (regular outbound) fee and the sender isn't aware of that yet, a window opens up momentarily where probing could reveal some information

Yes, that's a good point. If an intermediate node underpays the next node, here is what can happen:

- the payment is rejected: the attacker doesn't learn anything
- the payment is accepted: the attacker learns that the next node is *not* the final node. By looking at recent channel updates from the next node, the attacker may learn with a high probability who the subsequent node is.

I don't know if we should be concerned or not. We really don't want intermediate nodes to be able to learn that information. The only way I see to fully fix this would be to add a `required_incoming_htlc_amt` field as discussed in other comments. However, the conditions for this probing attack won't appear often in practice, as the payment needs to match a decrease in fees for one channel in the route. Also if we think something like [fee ratecards](https://lists.linuxfoundation.org/pipermail/lightning-dev/2022-September/003685.html) is the future, this will even decrease the probability that this probing attack succeeds.

> One thing to note is that there may be an additional cost for senders to keeping the path tight. If there is an intermediate node with a high min_htlc, the sender may not want to carry over that high amount to all channels further downstream. For those channels a proportional fee may be charged, in which case it would be cheaper for the sender to reduce amt_to_fwd right after the min_htlc has been passed. If the sender decides to do so, it becomes possible again to identify a node as being non-final.

I agree in theory, but I'm not sure it actually matters in practice. You would need to be using a route with a very high `min_htlc` in the middle **and** very high proportional fees after that node for this issue to cost you meaningful fees.

Even if those conditions do happen, the prober only learns that the next node is **not** the final node, which is much less worrying that learning who the final node is, so I think this is an acceptable probing vector (but maybe I'm wrong!).

Reply:
I agree that most of this may be theoretical. On the other hand though, if we would let the per-hop payload hmac also cover the required incoming htlc amount, all of these cases would be fixed for "free".

Reply:
> if we would let the per-hop payload hmac also cover the required incoming htlc amount

That would be an alternative for adding a `required_incoming_htlc_amt` field, right? But that removes all possibility of giving some leeway for intermediate nodes to decide to forward more to pay an inbound fee for example.

Reply:
Yes, a zero bytes alternative indeed for `required_incoming_htlc_amt` indeed. Why would an intermediate need leeway to pay for example an inbound fee? I'd think this is all decided upfront by the sender.

The feature bit is required indeed, but does the whole path need to support this? I thought the sender can calculate the hmac for each hop and include the required incoming amt only if the feature bit is present.

Reply:
Starting to magically include more data that's not be present in the HMAC sounds like something that would make similar changes harder later. 

What's bad about just adding an extra (optional) field `require_incoming_amount` and as more nodes start checking this, probing gets harder and harder. I still thinks this is the cleanest way of ensuring intermediate nodes can't do anything other than what the sender tells them to, which sounds like it would fix most of these probing vectors (maybe even some we haven't thought of).

Should also be totally backwards compatible IIUC.

I agree with t-bast that this would give intermediate nodes less leeway to overpay (for some reason), but as Joost is saying, shouldn't senders really be in charge of deciding that?

Reply:
>as an intermediate node or a final node, if I receive more money than expected, I will take it

Another way to look at this is that failing could lead to more money on the longer term. By failing you're teaching your predecessor that there is no point in trying to steal anything and next time they might just pay the instructed (higher) amount?